### DISCUSSION TITLE
Configurable unsorted-required-namespaces

### DISCUSSION THREAD

**Discussion by reutsharabani** [clj-kondo/clj-kondo#1583] [Ideas]
What are your thoughts about allowing configuration (probably a key-function just like `clojure.core/sort-by` has) that controls the sort order in unsorted-required-namespaces linter?

I think it raises several questions:
1. Do linters have linter-specific configuration today?
2. How will this function be exposed to kondo? I know hooks uses `.clj-kondo/hooks/...`

How much of an effort is it to support it? I took a (brief) look on [unsorted-required-imports](https://github.com/clj-kondo/clj-kondo/blob/9ec1d96b51f466dd0c3fae4cacce3087bbde7209/src/clj_kondo/impl/namespace.clj#L52) and it seems to use `compare` recursively so maybe a different method will be less of an effort?

Thanks!

**Comment by borkdude** [comment]
@reutsharabani 
1. yes, a lot: https://github.com/clj-kondo/clj-kondo/blob/master/doc/linters.md
2. such a function could be executed using SCI. Probably such a function would receive a map with the raw rewrite-clj node and possibly more, similar to hooks. But if we're going there, it's already starting to look a lot like a normal hook where you have more flexibility like you're doing it right now.
3. Also need to think about (spliced and non-spliced) reader conditionals. Currently these are ignored by clj-kondo as there is no defined sorting for those.

**Comment by reutsharabani** [comment]
It's worth adding to the discussion that writing a hook for our custom use was very easy.

**Comment by ericdallo** [comment]
If I may ask, what kind of different sorting do you seek? It may help add the problem you want to solve at your company, so we can think how clj-kondo can be improved to solve the issue, WDYT?
Just a little example may be enough I think

**Comment by reutsharabani** [comment]
@ericdallo an example would be to require internal libraries last.

I never used anything other than the liters' defaults in any company I worked (except my current one). So I don't think its a common use case.

But I do think supplying a key-function/comparator and basically configuring only the relation between required namespaces (and not writing the entire linter) is very flexible. And since it can be opt-in (default should be the current way) - I think it improves the linter overall.

**Comment by borkdude** [comment]
What I'm missing so far is the alternatives considered:

E.g. one could introduce a grouping construct so each group should be sorted (using the default linter). E.g.:

``` clojure
(ns foo
  (:require [external.lib1]
            [external.lib2]
            ;; internal libs
            [internal.lib1]
            [internal.lib2]
            ))
```

What is also a concern: the linter can warn on custom sorting, but tooling should be able to fix it too then probably: clojure-lsp, clj-refactor, etc.
